### PR TITLE
Convenience functions for vals

### PR DESCRIPTION
**Pull request by bramtayl**

### CODE REVIEW

File: base/essentials.jl [hunk 1]
```diff
@@ -601,6 +601,20 @@ end
 
 Val(x) = (@_pure_meta; Val{x}())
 
+eltype(::Val{T}) where T = T
```
Comment (ararslan) [inline base/essentials.jl]: This isn't correct. `T` can be `Array{S,N}`, in which case `eltype(Val{Array{S,N}})` should be `S`. I'd instead make this `eltype(::Val{T}) where {T} = eltype(T)`.

Comment (bramtayl) [reply base/essentials.jl]: Hmm maybe eltype isn't the right name for what I want. I just want a function that will turn `Val{:a}()` into `:a`

File: base/essentials.jl [hunk 2]
```diff
@@ -601,6 +601,20 @@ end
...
+
+(&)(v1::Val{true}, v2::Val{true}) = Val{true}()
+(&)(v1::Val, v2::Val) = Val{false}()
```
Comment (TotalVerb) [inline base/essentials.jl]: these methods seem to cover too many possibilities; I'd suggest to just write out all four combinations

**Review by vtjnash — CHANGES_REQUESTED** [review]
`Val` shouldn't be used with booleans or places where you want to do computations.

**Comment by bramtayl** [comment]
Ive been using value algebra for significant performance gains in places like JuliennedArrays

File: base/essentials.jl (hunk 1 above)
Comment (GregPlowman) [reply base/essentials.jl]: However, might be consistent with this:
```
Julia-0.7.0-DEV> x = Vector{Array{Int,3}}()
0-element Array{Array{Int64,3},1}

Julia-0.7.0-DEV> eltype(x)
Array{Int64,3}
```

**Comment by bramtayl** [comment]
> Val shouldn't be used with booleans or places where you want to do computations.

Why

**Comment by vtjnash** [comment]
See previous attempts at doing this (#22518), the manual (https://docs.julialang.org/en/stable/manual/performance-tips/#Types-with-values-as-parameters-1), the equivalent discussions on lifting Nullablility, and discourse.

**Comment by bramtayl** [comment]
Sounds as if there is indeed a specific use case:

```
This might be worthwhile when the following are true:

> * You require CPU-intensive processing on each Car, and it becomes vastly more efficient if you know the Make and Model at compile time.
> * You have homogenous lists of the same type of Car to process, so that you can store them all in an Array{Car{:Honda,:Accord},N}.

**Comment by vtjnash** [comment]
No, that is saying the opposite: Val is useful for cases where you don't need to do arithmetic on it.

**Comment by bramtayl** [comment]
I'm not using this these functions to do arithmetic though...

**Comment by bramtayl** [comment]
If by arithmetic you mean repeated functions on different values. Instead, I'm doing the same function on the exact same list of values repeatedly.

File: base/essentials.jl (hunk 1 above)
Comment (JeffBezanson) [reply base/essentials.jl]: `Val` isn't a collection or the type of a collection so `eltype` is not appropriate here.

Comment (bramtayl) [reply base/essentials.jl]: I mean I could just leave it out this PR, or choose another name? `devalue`?

**Comment by JeffBezanson** [comment]
I agree with @vtjnash that these cases should be handled by compiler constant folding instead of manually using `Val`. In particular `ifelse` shouldn't be necessary; if you have `Val{x}` then `ifelse(x, ...)` should already get constant folded.

**Comment by bramtayl** [comment]
It doesn't seem to be working though?

```
devalue(x::Val{T}) where T = T

Base.ifelse(switch::Val, new, old) = ifelse(devalue(switch), new, old)

tuple_mismatch_error(t...) = error("Mismatched tuples: leftovers $(t...)")

Base.getindex(into::Tuple{}, index::Tuple{}) = ()
Base.getindex(into::Tuple, index::Tuple{}) = tuple_mismatch_error(into)
Base.getindex(into::Tuple{}, index::Tuple) = tuple_mismatch_error(index)
Base.getindex(into::Tuple, index::Tuple) = begin
    next = getindex(tail(into), tail(index))
    ifelse(first(index), (first(into), next...), next)
end

@code_warntype getindex((1, "a", 3), (Val{true}(), Val{false}(), Val{true}()))
```

**Comment by bramtayl** [comment]
My follow-up pull request was going to be a type stable `getindex` and `setindex` for tuples.

**Comment by bramtayl** [comment]
Both of which will be needed for a type-stable mapslices.

**Comment by Cody-G** [comment]
I've also worked on this before, and I had trouble finding a way to get constant folding to work in the way that's necessary for a package like @bramtayl's [JuliennedArrays](https://github.com/bramtayl/JuliennedArrays.jl) (which already works well).  I've also read the manual and any relevant issues/discourse discussions that I can find, and I'm very interested to learn what is the proper way to do this with Julia v0.7 master.

**Comment by bramtayl** [comment]
Constant folding seems very incomplete. Here's another example where constant folding is not working as expected so value wrapping is necessary:

```
value_wrap(f) = (x, args...) -> Val(f(devalue(x), args...))

fieldtypes(any) = fieldtypes(Val(any))
function fieldtypes(val_type::Val)
    Base.@pure inner_function(i) = value_wrap(fieldtype)(val_type, i)
    ntuple(inner_function, value_wrap(nfields)(val_type))
end

# doesn't constant fold
fieldtypes(val_type::Val) =
    ntuple(nfields(val_type)) do
        fieldtype(val_type, i)
    end
```

**Comment by bramtayl** [comment]
Am I correct in my understanding that constant propagation cannot survive passing through functions not known to the compiler as pure? If so, it severely limits the usefulness of constant propagation.

**Comment by bramtayl** [comment]
My understanding is that this is a non-issue with the now aggressive constant propogation

**Pull request closed without merging**